Add AppleScript Support

SKaplanOfficial opened the pull request:
I've added AppleScript equivalents for all of the Shortcuts, plus support for the AppleScript standard dictionary and text suite. The full supported terminology can be viewed via Script Editor's File->Open Dictionary menu or in `MarkEditMac/Resources/MarkEdit.sdef`

I haven't done much Swift development, so I probably used odd approaches in some places. Let me know if I need to rework anything. Also, I'm not sure why Info.plist got all rearranged — is there something I can do to fix that?

Here's a test script demonstrating what's currently supported:

```applescript
tell application "MarkEdit"
	-- Create new documents
	set doc1 to make new document with properties {name:"test1.md"}
	set doc2 to make new document with properties {name:"test2.md"}
	set doc3 to make new document with properties {name:"test3.md"}

	-- Modify file content
	delay 0.5
	set source of doc1 to "# Hello, world!

The *quick* brown fox **jumps** over the lazy *dog*.

- Item 1
- Item 2
- Item 3"
	set source of doc2 to "# Testing..."
	set source of doc3 to source of doc2

	set targetDoc to first document whose source contains "dog"

	-- Get rich text details
	set italicWords to words of formatted text of targetDoc whose font contains "Oblique"
	if italicWords is {"quick", "dog"} then
		set targetWindow to first window where its document is targetDoc
		set index of targetWindow to 1

		tell targetDoc
			-- Run JavaScript on documents
			evaluate JavaScript "document.body.style.zoom = '200%'"

			-- Save documents
			set outputPath to POSIX file (POSIX path of (path to downloads folder) & "test1.md")
			close saving in outputPath
		end tell
	end if

	close doc2 without saving
	close doc3 without saving
end tell
```

Known issues:

- Specifying the `source` property when using the `make` command raises an error. This is because setting `source` relies on there being an existing webview to call `replaceText` on. Current workaround: Use a momentary delay after `make`, then use `set source to ...`.
- The `color` property of rich text elements is always `{0, 0, 0}`. This is expected since there is no color information included in the document's `source`. A future improvement could be to process the markdown source and add colors based on the current theme.

Closes #854.

cyanzhong commented:
Thank you for the contribution, Stephen!

Regarding the plist change, it looks strange to me too. Can you share the actual changes needed here?

Is it just:

```xml
<key>NSAppleScriptEnabled</key>
<true/>
<key>OSAScriptingDefinition</key>
<string>MarkEdit.sdef</string>
```

cyanzhong commented on MarkEdit.xcodeproj/project.pbxproj:
```diff
@@ -971,7 +983,7 @@
 				INFOPLIST_KEY_NSHumanReadableCopyright = "";
 				INFOPLIST_KEY_NSMainStoryboardFile = Main;
-				INFOPLIST_KEY_NSPrincipalClass = "$(PRODUCT_NAME).Application";
+				INFOPLIST_KEY_NSPrincipalClass = MarkEdit.Application;
```
Is using static name here required?

cyanzhong commented on MarkEditMac/Sources/Scripting/EditorDocument+Scripting.swift:
```diff
@@ -0,0 +1,174 @@
+    get {
+      return self.stringValue
+    }
+    set(newValue) {
```
nit: `(newValue)` can be omitted.

cyanzhong commented on MarkEditMac/Sources/Scripting/EditorDocument+Scripting.swift:
```diff
@@ -0,0 +1,174 @@
+import Carbon
+import MarkEditKit
+
+extension EditorDocument {
```
Can we name `source` and `formattedText` to be more scripting specific? Since we are creating extension that can be used in the entire app.

cyanzhong replied:
If these are something to be used in the script, it's fine to keep them clean (or follow some conventions).

SKaplanOfficial replied:
Sure, would prefixing them with `scripting` be enough? E.g. `scriptingSource`?

cyanzhong commented on MarkEditMac/Sources/Scripting/EditorDocument+Scripting.swift:
```diff
@@ -0,0 +1,174 @@
+  /// Executes a user-provided JS script in the document's webview.
+  @objc func handleEvaluateCommand(_ command: NSScriptCommand) -> Any? {
+    guard let inputString = command.evaluatedArguments?["script"] as? String else {
+      let scriptError = ScriptError.missingInput(message: "No JavaScript script provided")
```
Is this user-facing string? I can localize it later if so.

SKaplanOfficial replied:
It would get presented to users as an AppleScript error, but I think it would only occur if the incoming event data gets corrupted, which is very rare. Just omitting the script parameter or supplying a non-text value gets handled by AppleScript before sending any messages to the application.

Maybe it's not worth keeping the message? The error number would still appear anyway.

cyanzhong commented on MarkEditMac/Sources/Scripting/EditorDocument+Scripting.swift:
```diff
@@ -0,0 +1,174 @@
+      return nil
+    }
+
+    guard let targetEditor = self.windowControllers.first?.contentViewController as? EditorViewController else {
```
nit: code convention wise, I prefer to hide `self.` if we can. I can do a cleanup later.

cyanzhong commented on MarkEditMac/Sources/Scripting/NSColor+Scripting.swift:
```diff
@@ -0,0 +1,48 @@
+  static let RGBColorCoefficient: CGFloat = 65535
+
+  /// Unpacks incoming color descriptors into NSColor objects.
+  @objc func scriptRGBColor(with descriptor: NSAppleEventDescriptor) -> NSColor? {
```
Just curious, are these functions called by scripts? Don't see a caller in the code.

SKaplanOfficial replied:
The get called automatically by the Apple Event Manager when users send/receive color objects via AppleScript. They're normally automatically generated at runtime, but that's skipped for colors due to how the text suite defines them.

cyanzhong commented on MarkEditMac/Sources/Scripting/NSColor+Scripting.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    return NSColor(
+      calibratedRed: CGFloat(rgbColor.red) / Self.RGBColorCoefficient,
```
I am not an expert of this, just making sure we are using correct initializer, is `calibrated` needed?

SKaplanOfficial replied:
Not sure, the only clear reference for those methods is [iterm2](https://gitlab.com/gnachman/iterm2/-/blob/v3.3.8beta4/sources/NSColor+Scripting.m). I think the unpack method can use any NSColor init method, but I'd need to check.

cyanzhong commented on MarkEditMac/Sources/Scripting/EditorDocument+Scripting.swift:
```diff
@@ -0,0 +1,174 @@
+//  Created by Stephen Kaplan on 4/2/25.
+//
+
+import Carbon
```
This is deprecated, does `Foundation` work?

SKaplanOfficial replied:
No, `keyASUserRecordFields` is only defined under Carbon, but I don't know a way to construct valid NSAppleEventDescriptors for dictionaries without that key. Apple [still uses it](https://developer.apple.com/documentation/professional-video-applications/supporting-conversions-with-scripting-class-extensions) in non-archived documentation.

I can use the raw OSType int value instead, which I guess it better.

cyanzhong commented on MarkEditMac/Sources/Scripting/ScriptErrors.swift:
```diff
@@ -0,0 +1,42 @@
+    case .noActiveEditor:
+      return -3
+    case .jsEvalError:
+      return -1702
```
Why `-1702`?

SKaplanOfficial replied:
Good catch, those values were placeholders. Was going to use the OSType of the AS event class instead.

cyanzhong commented:
Thanks again, Stephen. The change overall LGTM, I just left some comments.

I have one more general question about the change, since MarkEdit only handles plain text, why is rich text support needed?

SKaplanOfficial commented:
> Thank you for the contribution, Stephen!

No problem, it was a fun challenge to tackle!

> Is it just:
> 
> ```
> <key>NSAppleScriptEnabled</key>
> <true/>
> <key>OSAScriptingDefinition</key>
> <string>MarkEdit.sdef</string>
> ```

I believe so, but I'll have to make sure. Apple Events are very picky. Won't get the changes done until tomorrow, but I'll respond to your other comments now.

> I couldn't think of a better way either. I used similar delays in the app to handle a similar case (see `newFileFromClipboard`).

Hm, I was thinking about looking into `@Observable`. If we separate the stringValue and other properties derived from the webview bridge into an observable class, then both EditorDocument and EditorViewController could react to changes more easily, maybe. Could be a big change, so definitely out of scope for this.

> I have one more general question about the change, since MarkEdit only handles plain text, why is rich text support needed?

The text suite (which includes the 'rich text' AppleScript class) is the recommended way to access individual words, characters, etc. from AppleScript since repeating over elements is much, much faster in Swift/ObjC, so that's the reason for the additional property in general.

Then it just seemed reasonable to use Swift's markdown support in NSAttributedString so you can detect when text is italic/bold without including the surrounding symbols in the "words" subproperty.

cyanzhong commented:
I would like to update the wiki after we merge the change and looking for these example:

- Getting the content of an opened document
- Replacing the selection with some text

Could you please kindly provide some example scripts? Thanks!

cyanzhong commented:
For the plist change, did you use Xcode to generate the change, or did you modify the plist file directly?

Xcode may re-generate the entire file which may lead to unwanted changes.

SKaplanOfficial commented:
Made those changes, and added the error message strings to AppResources.Localized.Scripting.

>     * Replacing the selection with some text

Knew I was forgetting something. Added the `selection` property in the latest commit.

I made some examples here: https://gist.github.com/SKaplanOfficial/9637cd69e2b21b1749521aafa1f71c95



> For the plist change, did you use Xcode to generate the change, or did you modify the plist file directly?

I used Xcode, so that explains it! I've reverted those changes and manually added the ones for AppleScript.

cyanzhong commented on MarkEditMac/Resources/MarkEdit.sdef:
```diff
@@ -0,0 +1,98 @@
+
+    <class name="document" code="docu" description="A MarkEdit document." plural="documents" inherits="document">
+      <cocoa class="MarkEdit.EditorDocument"/>
+      <property name="source" code="mdsc" type="text" access="rw" description="The raw markdown source of the document.">
```
Does this mean the variable name in scripts is `source`, which uses `scriptingSource` in code?

SKaplanOfficial replied:
Yes, exactly.

cyanzhong commented on README.md:
```diff
@@ -23,7 +23,7 @@ MarkEdit is a free and **open-source** Markdown editor, for macOS. It's just lik
 - Fast: edits 10 MB files easily
 - Lightweight: installer size is about 3 MB
-- Extensible: seamless Shortcuts integration
+- Extensible: seamless integration with Shortcuts and AppleScript
```
👍

cyanzhong commented on MarkEditMac/Sources/Scripting/EditorDocument+Scripting.swift:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    // Preserve newlines for better error reporting
+    let jsText = inputString.replacingOccurrences(of: "\\n", with: "\\\\n")
```
@SKaplanOfficial this quite interesting to me, does this really work?

If we are replacing `\n` to `\\n` in the scripts, should it be `inputString.replacingOccurrences(of: "\n", with: "\\n")`?

I don't quite get why we need double backslashes, and why we need to replace them at all.
